use-case-guides/workday-request-time-off/.codeblocks/block_3.py

```diff
@@ -1,91 +1,245 @@
-from fastapi import FastAPI, HTTPException
+from fastapi import FastAPI, HTTPException, Path
+from fastapi.responses import JSONResponse
 import requests
+from datetime import datetime, timedelta
+from pydantic import BaseModel
+from typing import List, Optional
+import httpx
 
-CANONICAL_TIME_OFF_PLAN_MAP = {
-  "Time off (USA)": "67839087362157384953",
-  "Wellness Days (USA)": "802394093284924538920",
-  ...
-}
-
+# Create an instance of the FastAPI application
 app = FastAPI()
 
-# Function to get access token
-def get_access_token():
-    workday_instance_id = "your_workday_instance_id"
-    url = "https://wd2-impl-services1.workday.com/ccx/oauth2/{workday_instance_id}/token"
+# Define data models using Pydantic
+
+class AuthResponse(BaseModel):
+    """Model for authentication response"""
+    access_token: str
+    token_type: str
+    refresh_token: str
+
+class WorkerQueryRequest(BaseModel):
+    """Model for worker query request"""
+    query: str
+
+class Worker(BaseModel):
+    """Model representing a worker's details"""
+    workdayID: str
+    fullName: str
+    email_PrimaryWork: str
+    position: dict
+    allEligibleTimeOffsForWorker: List[dict]
+
+class WorkerQueryResponse(BaseModel):
+    """Model for worker query response"""
+    total: int
+    data: List[Worker]
+
+class TimeOffDay(BaseModel):
+    """Model for a single day of time off"""
+    date: str
+    dailyQuantity: int
+    comment: str
+    timeOffType: dict
+    position: dict
+
+class TimeOffRequest(BaseModel):
+    """Model for a time off request"""
+    start_date: str
+    time_off_type_id: str
+    end_date: Optional[str] = None
+    comment: Optional[str] = None
+
+class TimeOffResponseDay(BaseModel):
+    """Model for a single day in the time off response"""
+    id: str
+    descriptor: str
+    dailyQuantity: str
+    comment: str
+    position: dict
+    date: str
+    timeOffType: dict
+
+class TimeOffResponse(BaseModel):
+    """Model for the time off response"""
+    businessProcessParameters: dict
+    days: List[TimeOffResponseDay]
+
+# Constants for API URLs and authentication. Please store username and password securely inside environment variables.
+AUTH_URL = "https://{tenant}.workday.com/ccx/oauth2/{org}/token"  # Replace with your Workday org id and tenant
+BASIC_AUTH_HEADER = "Basic REDACTED"  # Replace with your base64 encoded username:password. You can use tools like https://mixedanalytics.com/tools/basic-authentication-generator/ to generate the base64 encoded string.
+REFRESH_TOKEN = "REDACTED"  # Replace with your refresh token
```

Comment:
This is suggesting that developers hard code their credentials into the iPaaS code. I don't think that passes our security bar and Tim commented we need to be strict about this if we're going to publish guidance like this.